### ISSUE TITLE
SIGSEGV on FreeBSD: src/flb_scheduler.c:52 errno=0

### ISSUE THREAD

**Issue by nomadlogic** [fluent/fluent-bit#2747]
## Bug Report

**Describe the bug**
When running fluent-bit on FreeBSD systems using the random input and stdout output methods I get a SIGSEGV and crash of the fluent-bit process.

**To Reproduce**
Install fluent-bit on FreeBSD-12.2 or FreeBSD-CURRENT, then run fluent-bit like so:
`fluent-bit -i random -o stdout`

Output of running fluent-bit to demonstrate this issue:
```
$ fluent-bit -i random -o stdout
Fluent Bit v1.6.3
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/11/04 20:06:30] [ info] [engine] started (pid=8807)
[2020/11/04 20:06:30] [ info] [storage] version=1.0.6, initializing...
[2020/11/04 20:06:30] [ info] [storage] in-memory
[2020/11/04 20:06:30] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/11/04 20:06:30] [ info] [sp] stream processor started
[2020/11/04 20:06:31] [error] [src/flb_scheduler.c:52 errno=0] No error: 0
[2020/11/04 20:06:32] [error] [src/flb_scheduler.c:52 errno=0] No error: 0
[2020/11/04 20:06:33] [error] [src/flb_scheduler.c:52 errno=0] No error: 0
[2020/11/04 20:06:34] [error] [src/flb_scheduler.c:52 errno=0] No error: 0
[2020/11/04 20:06:35] [engine] caught signal (SIGSEGV)
ERROR: no stack trace because unwind library not available (0)Abort trap (core dumped)

```

**Expected behavior**
fluent-bit should run as expected according to documentation and not exit on startup.

**Your Environment**

* Version used: 1.6.3
* Operating System and version: FreeBSD-12.2/FreeBSD-CURRENT AMD64
* Output of `fluent-bit --help` listing built in options:

```
$ fluent-bit --help
Usage: fluent-bit [OPTION]

<snip>

Internal
 Event Loop  = kqueue
 Build Flags =  FLB_HAVE_PARSER FLB_HAVE_RECORD_ACCESSOR FLB_HAVE_STREAM_PROCESSOR JSMN_PARENT_LINKS JSMN_STRICT FLB_HAVE_TLS FLB_HAVE_AWS FLB_HAVE_SIGNV4 FLB_HAVE_SQLDB FLB_HAVE_TRACE FLB_HAVE_FORK FLB_HAVE_TIMESPEC_GET FLB_HAVE_GMTOFF FLB_HAVE_UNIX_SOCKET FLB_HAVE_PROXY_GO FLB_HAVE_LIBBACKTRACE FLB_HAVE_REGEX FLB_HAVE_UTF8_ENCODER FLB_HAVE_LUAJIT FLB_HAVE_C_TLS FLB_HAVE_ACCEPT4
```

**Additional context**
There is an official FreeBSD port which has worked in the past.  I have opened a FreeBSD bugreport viewable here with our port maintainer:
https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=250825

As mentioned in that ticket running this process through truss seems to indicate a missing file error, but I am not certain that is the root cause:

```
34233: openat(AT_FDCWD,"/dev/urandom",O_RDONLY,00) = 24 (0x18)
34233: read(24,"\M-eH\M-t\M-X\^U\M-z\M-.\M^D",8) = 8 (0x8)
34233: close(24)				 = 0 (0x0)
34233: getpid()					 = 34233 (0x85b9)
34233: kevent(17,0x0,0,{ 23,EVFILT_TIMER,EV_CLEAR,0,0x1,0x41609150 },256,0x0) = 1 (0x1)
34233: read(23,0x7fffdfffcea0,8)		 = 0 (0x0)
34233: fstatat(AT_FDCWD,"/usr/share/nls/C/libc.cat",0x7fffdfffcbb0,0x0) ERR#2 'No such file or directory'
34233: fstatat(AT_FDCWD,"/usr/share/nls/libc/C",0x7fffdfffcbb0,0x0) ERR#2 'No such file or directory'
34233: fstatat(AT_FDCWD,"/usr/local/share/nls/C/libc.cat",0x7fffdfffcbb0,0x0) ERR#2 'No such file or directory'
34233: fstatat(AT_FDCWD,"/usr/local/share/nls/libc/C",0x7fffdfffcbb0,0x0) ERR#2 'No such file or directory'
34233: ioctl(1,TIOCGETA,0x7fffdfffbb70)		 = 0 (0x0)
34233: write(12,"d\0\0\0\0\0\0\0\^[[1m[\^[[0m2020/11/04 20:15:59\^[[1m]\^[[0m [\^[[91merror\^[[0m] [src/flb_scheduler.c:52 errno=0] No error: 0\n\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,4096) = 4096 (0x1000)
34233: kevent(8,0x0,0,{ 11,EVFILT_READ,0x0,0,0x1000,0x41609000 },16,0x0) = 1 (0x1)
34233: read(11,"d\0\0\0\0\0\0\0\^[[1m[\^[[0m2020/11/04 20:15:59\^[[1m]\^[[0m [\^[[91merror\^[[0m] [src/flb_scheduler.c:52 errno=0] No error: 0\n\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"...,4096) = 4096 (0x1000)
[2020/11/04 20:15:59] [error] [src/flb_scheduler.c:52 errno=0] No error: 0
34233: write(2,"\^[[1m[\^[[0m2020/11/04 20:15:59\^[[1m]\^[[0m [\^[[91merror\^[[0m] [src/flb_scheduler.c:52 errno=0] No error: 0\n",100) = 100 (0x64)
34233: nanosleep({ 1.000000000 })		 = 0 (0x0)
```

I am happy to provide any additional information that will help debug this.

**Comment by nomadlogic** [comment]
I have also obtained a backtrace from lldb and have a core file available for analysis:

```
$ lldb -c ./fluent-bit.core -- fluent-bit
(lldb) target create "fluent-bit" --core "./fluent-bit.core"
Core file '/home/pete/git/fluent-bit/build/bin/fluent-bit.core' (x86_64) was loaded.

(lldb) thread backtrace all
This version of LLDB has no plugin for the mipsassem language. Inspection of frame variables will be limited.
* thread #1, name = 'fluent-bit', stop reason = signal SIGABRT
  * frame #0: 0x000000004087100a libc.so.7`__sys_thr_kill at thr_kill.S:4
    frame #1: 0x00000000407e6c84 libc.so.7`__raise(s=6) at raise.c:52:10
    frame #2: 0x000000004089a5d9 libc.so.7`abort at abort.c:67:8
    frame #3: 0x000000000034a7a8 fluent-bit`flb_signal_handler(signal=11) at fluent-bit.c:418:9
    frame #4: 0x00000000406d1c20 libthr.so.3`handle_signal(actp=0x00007fffdfffc600, sig=11, info=0x00007fffdfffc9f0, ucp=0x00007fffdfffc680) at thr_sig.c:303:3
    frame #5: 0x00000000406d11ef libthr.so.3`thr_sighandler(sig=11, info=0x00007fffdfffc9f0, _ucp=0x00007fffdfffc680) at thr_sig.c:246:2
    frame #6: 0x00007fffffffe193
    frame #7: 0x000000000036fe0c fluent-bit`tasks_start [inlined] output_params_set(th=0x00000000416091c0, data=0x00000000416648c0, bytes=160, tag="random.0", tag_len=8, i_ins=0x0000000040e58000, out_plugin=0x0000000040e2dfc0, out_context=0x00000000416051e0, config=0x0000000040e19180) at flb_output.h:429:5
    frame #8: 0x000000000036fd14 fluent-bit`tasks_start [inlined] flb_output_thread(task=0x00000000416410a0, i_ins=0x0000000040e58000, o_ins=0x0000000040e5b000, config=0x0000000040e19180, buf=0x00000000416648c0, size=160, tag="random.0", tag_len=8) at flb_output.h:522
    frame #9: 0x000000000036fab5 fluent-bit`tasks_start(in=0x0000000040e58000, config=0x0000000040e19180) at flb_engine_dispatch.c:190
    frame #10: 0x000000000036f6fd fluent-bit`flb_engine_dispatch(id=0, in=0x0000000040e58000, config=0x0000000040e19180) at flb_engine_dispatch.c:293:5
    frame #11: 0x000000000036ce10 fluent-bit`flb_engine_flush(config=0x0000000040e19180, in_force=0x0000000000000000) at flb_engine.c:85:9
    frame #12: 0x000000000036d43a fluent-bit`flb_engine_start [inlined] flb_engine_handle_event(fd=20, mask=1, config=0x0000000040e19180) at flb_engine.c:292:13
    frame #13: 0x000000000036d3e6 fluent-bit`flb_engine_start(config=0x0000000040e19180) at flb_engine.c:559
    frame #14: 0x00000000003564ec fluent-bit`flb_lib_worker(data=0x0000000040e19180) at flb_lib.c:488:11
    frame #15: 0x00000000406c87db libthr.so.3`thread_start(curthread=0x0000000040e12500) at thr_create.c:292:16
  thread #2, name = 'fluent-bit', stop reason = signal SIGABRT
    frame #0: 0x00000000408721ca libc.so.7`__sys_nanosleep at _nanosleep.S:4
    frame #1: 0x00000000406d3c3c libthr.so.3`__thr_nanosleep(time_to_sleep=<unavailable>, time_remaining=<unavailable>) at thr_syscalls.c:289:8
    frame #2: 0x00000000407e96eb libc.so.7`__sleep(seconds=1) at sleep.c:62:6
    frame #3: 0x0000000000356548 fluent-bit`flb_loop(ctx=0x0000000040e18000) at flb_lib.c:552:9
    frame #4: 0x0000000000348e1d fluent-bit`flb_main(argc=5, argv=0x00007fffffffd788) at fluent-bit.c:1034:5
    frame #5: 0x000000000034a492 fluent-bit`main(argc=5, argv=0x00007fffffffd788) at fluent-bit.c:1045:12
    frame #6: 0x00000000003485d0 fluent-bit`_start(ap=<unavailable>, cleanup=<unavailable>) at crt1_c.c:75:7
  thread #3, name = 'fluent-bit', stop reason = signal SIGABRT
    frame #0: 0x000000004087218a libc.so.7`__sys_kevent at _kevent.S:4
    frame #1: 0x00000000406d4303 libthr.so.3`__thr_kevent(kq=8, changelist=0x0000000000000000, nchanges=0, eventlist=0x0000000041602000, nevents=16, timeout=<unavailable>) at thr_syscalls.c:400:8
    frame #2: 0x0000000000662e6f fluent-bit`cb_grep_filter(data=0x0000000041601000, bytes=1096818688, tag=0x0000000000000000, tag_len=0, out_buf=0x00007fffdfdfbf10, out_size=0x0000000040e12500, f_ins=0x0000000041600000, context=0x0000000041605000, config=0x00007fffdfdfbf90) at grep.c:241:15
    frame #3: 0x0000000000662e15 fluent-bit`cb_grep_filter(data=0x0000000041600000, bytes=6696469, tag="\x90\xbf\xdf\xdf\xff\U0000007f", tag_len=0, out_buf=0x0000000041601000, out_size=0x0000000041602000, f_ins=0x00007fffdfdfbf90, context=0x0000000000356781, config=0xffffffff00000000) at grep.c:231:23
    frame #4: 0x0000000000357027 fluent-bit`log_worker_collector(data=0x0000000041600000) at flb_log.c:126:9
    frame #5: 0x00000000003a50e6 fluent-bit`step_callback(data=0x0000000041609070) at flb_worker.c:51:5
    frame #6: 0x00000000406c87db libthr.so.3`thread_start(curthread=0x0000000041611000) at thr_create.c:292:16
(lldb) 
```

**Comment by ypankov** [comment]
I have updated the FreeBSD PR with a bit more information.

TL;DR clang/lld don't have special meaning for `.text#` section as apparently gcc/gnu ld do, so having `text_session` defined to `__attribute__((section(".text#")))` makes the former put the `co_swap_function` in `.text#` section not having the execute permission, hence trying to execute it results in SIGSEGV.

The possible fix is to define `text_section` to be `__attribute__((section(".text")))` for FreeBSD/clang/lld in lib/flb_libco/settings.h.

**Comment by edsiper** [comment]
regarding compiler and libco:

I think the proper fix will be to upgrade the original libco dependency which has been changed a lot recently:

https://github.com/higan-emu/higan/tree/master/libco

this fix will be merged during 1.7 dev cycle.

**Comment by ypankov** [comment]
> regarding compiler and libco:
> 
> I think the proper fix will be to upgrade the original libco dependency which has been changed a lot recently:
> 
> https://github.com/higan-emu/higan/tree/master/libco
> 
> this fix will be merged during 1.7 dev cycle.

Looking at the code, it still has same assembly (https://github.com/higan-emu/higan/blob/2cf2ed6406f957372b225a4584b9650fafef693a/libco/amd64.c#L78) preceded by the same section definition (https://github.com/higan-emu/higan/blob/2cf2ed6406f957372b225a4584b9650fafef693a/libco/amd64.c#L16) expanding into the same `.text#` (https://github.com/higan-emu/higan/blob/2cf2ed6406f957372b225a4584b9650fafef693a/libco/settings.h#L99).

**Comment by girgen** [comment]
pull request exists in https://github.com/fluent/fluent-bit/pull/2781

**Comment by orivej** [comment]
By default LLD puts read-only sections into a non-executable segment:
```
$ ld.lld --help | grep rosegment
  --no-rosegment          Do not put read-only non-executable sections in their own segment
  --rosegment             Put read-only non-executable sections in their own segment (default)
```
To circumvent this you could put something executable into the same section, e.g. `text_section void co_make_this_section_executable() {}`.

**Comment by orivej** [comment]
Oh, this issue has already been properly fixed in https://github.com/edsiper/flb_libco/pull/9 and https://github.com/fluent/fluent-bit/commit/dafd450c

**Comment by asdavidov** [comment]
It should be noted that those flb_libco fixes should be also merged into monkey http server code to fully mitigate this issue.

**Comment by nomadlogic** [comment]
Thanks everyone for the help and attention to this issue.  Closing!

**Comment by asdavidov** [comment]
This issue is still present in the code.
The /lib/monkey/deps/flb_libco/settings.h file is still not patched and contains code that could lead to SIGSEGV if it is built with clang compiler.
The flb_libco fixes should be also merged in monkey server code to fully mitigate this issue.
So I guess that this issue should be reopened.

**Comment by nomadlogic** [comment]
Apologies @asdavidov - I was confused about the open task of updating the code under `lib/monkey/deps/flb_libco/settings.h`. I am guessing the update applied here: https://github.com/fluent/fluent-bit/commit/dafd450c would also need to be applied?

**Comment by edsiper** [comment]
Fixes merged in 8fe5a131c6
